test/hotspot/jtreg/serviceability/jvmti/RedefineClasses/RedefineRetransform/RedefineRetransform.java

```diff
@@ -0,0 +1,256 @@
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ *
+ * @bug 7124710
+ *
+ * @requires vm.jvmti
+ * @modules java.base/jdk.internal.org.objectweb.asm
+ * @library /test/lib
+ *
+ * @comment main/othervm/native -Xlog:redefine*=trace -agentlib:RedefineRetransform RedefineRetransform
+ * @run main/othervm/native -agentlib:RedefineRetransform RedefineRetransform 1
+ * @run main/othervm/native -agentlib:RedefineRetransform RedefineRetransform 2
+ * @run main/othervm/native -agentlib:RedefineRetransform RedefineRetransform 3
+ * @run main/othervm/native -agentlib:RedefineRetransform RedefineRetransform 4
+ * @run main/othervm/native -agentlib:RedefineRetransform RedefineRetransform 5
+ * @run main/othervm/native -agentlib:RedefineRetransform RedefineRetransform 6
+ */
+
+import java.io.IOException;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+
+import jdk.internal.org.objectweb.asm.AnnotationVisitor;
+import jdk.internal.org.objectweb.asm.ClassReader;
+import jdk.internal.org.objectweb.asm.ClassVisitor;
+import jdk.internal.org.objectweb.asm.ClassWriter;
+import jdk.internal.org.objectweb.asm.Opcodes;
+import jdk.internal.org.objectweb.asm.Type;
+
+public class RedefineRetransform {
+    static {
+        System.loadLibrary("RedefineRetransform");
+    }
+
+    @Retention(RetentionPolicy.RUNTIME)
+    @interface ClassVersion {
+        int value();
+    }
+
+    // Use runtime-visible annotation to specify class version
+    @ClassVersion(0)
+    static class TestClass {
+        public TestClass() { }
+    }
+
+    // redefines testClass with classBytes, instrument with classLoadHookBytes (is != null)
+    private static native byte[] nRedefine(Class testClass, byte[] classBytes, byte[] classLoadHookBytes);
+    // retransforms testClass with classBytes (if != null)
+    private static native byte[] nRetransform(Class testClass, byte[] classBytes);
+
+    private static byte[] initialClassBytes;
+
+    private static class VersionScanner extends ClassVisitor {
```

Comment:
This class needs some explanation.
It is unclear how does it work.
Could you, please, add relevant comments where needed?
For instance, how are the class file bytes constructed and what role does the version play.

Reply:
Done.